In [10]:
import pandas as pd 
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.neighbors import KNeighborsClassifier
from sklearn.cluster import KMeans
from sklearn.metrics import accuracy_score, adjusted_rand_score
import numpy as np 
from sklearn.datasets import load_digits

In [2]:
data = load_digits()

In [3]:
X,y = data.data, data.target

In [4]:
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=0.2, random_state=42)

# ✅ k-NN Classification

In [5]:
knn_model = KNeighborsClassifier()
knn_model.fit(X_train, y_train)
y_pred_knn = knn_model.predict(X_test)

knn_accuracy_score = accuracy_score(y_test, y_pred_knn)
print(f"✅ k-NN Classification Accuracy: {knn_accuracy_score}")

✅ k-NN Classification Accuracy: 0.9861111111111112


# 📍 k-Means Clustering

In [17]:
# Standardize features for clustering
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X_train)

# Fit k-Means on training data
kmeans = KMeans(n_clusters=10, random_state=42, n_init=10)
kmeans.fit(X_scaled)

# Cluster labels
y_pred_kmeans = kmeans.labels_

# Evaluate clustering with ARI
kmeans_ari = adjusted_rand_score(y_train, y_pred_kmeans)

print(f"📍 k-Means Clustering Adjusted Rand Index (ARI): {kmeans_ari:.4f}")

📍 k-Means Clustering Adjusted Rand Index (ARI): 0.4651


### ✅ k-NN vs 📍 k-Means – Comparison

**Feature	k-NN (k-Nearest Neighbors)	--> k-Means Clustering**

* Type	    Supervised Learning	        --> Unsupervised Learning
* Uses Labels?	✅ Yes	                    ❌ No
* Goal	Classify data	                    Group data into clusters
* Accuracy Metric	Accuracy Score	        Adjusted Rand Index (ARI)
* Result	✅ 0.9861 Accuracy	            📍 0.4651 ARI

**🔍 Conclusion:**
k-NN outperforms k-Means in classification tasks due to its use of labeled data, while k-Means is better suited for exploratory data analysis or clustering when labels are unknown.